app/src/main/java/io/gnosis/kouban/ui/balances/BalancesWidgetProvider.kt

```diff
@@ -0,0 +1,97 @@
+package io.gnosis.kouban.ui.balances
+
+import android.app.PendingIntent
+import android.app.Service
+import android.appwidget.AppWidgetManager
+import android.appwidget.AppWidgetProvider
+import android.content.Context
+import android.content.Intent
+import android.os.IBinder
+import android.widget.RemoteViews
+import com.squareup.picasso.Picasso
+import io.gnosis.kouban.R
+import io.gnosis.kouban.core.managers.SafeAddressManager
+import io.gnosis.kouban.core.utils.setTransactionIcon
+import io.gnosis.kouban.data.repositories.SafeRepository
+import io.gnosis.kouban.data.utils.shiftedString
+import io.gnosis.kouban.ui.MainActivity
+import kotlinx.coroutines.runBlocking
+import org.koin.android.ext.android.inject
+
+
+class BalancesWidgetProvider : AppWidgetProvider() {
+
+    override fun onUpdate(context: Context?, appWidgetManager: AppWidgetManager?, appWidgetIds: IntArray?) {
+        // To prevent any ANR timeouts, we perform the update in a service
+        context?.startService(
+            Intent(context, UpdateService::class.java)
+                .putExtra(UpdateService.EXTRA_WIDGET_IDS, appWidgetIds)
+        )
+    }
+
+    class UpdateService : Service() {
+
+        private val addressManager: SafeAddressManager by inject()
+        private val prefs: BalancesWidgetPrefs by inject()
+        private val safeRepository: SafeRepository by inject()
+        private val picasso: Picasso by inject()
+
+        override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+
+            val appWidgetManager = AppWidgetManager.getInstance(this)
+            val appWidgetIds = intent?.getSerializableExtra(EXTRA_WIDGET_IDS) as IntArray
+
+            runBlocking {
+
+                val safe = addressManager.getSafeAddress()!!
+                val balances = safeRepository.loadTokenBalances(safe)
+
+                appWidgetIds?.forEach { appWidgetId ->
```

Comment:
I am a little confused, in some places we bave code to handle multiple different widget ids, but in others it defaults to one hardcodes id could this be unified or commented?

Reply:
yes that is how it should be in theory but you rimplementation doesn't do that as far as I can tell (see above) ... also could be ignored for now.